### PR TITLE
Support local cache on memcache store read multi

### LINKED ISSUE #31909: Memcache: Rails.cache.fetch_multi is slow by an order of magnitude than Rails.cache.read
### Steps to reproduce

Using the bug report template doesn't reproduce the issue for some reason:

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "rails", "~> 5.1.0"
  gem "dalli"
end

require "rack/test"
require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  config.session_store :cookie_store, key: "cookie_store_key"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/" => "test#index"
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def index
    cache = ActiveSupport::Cache.lookup_store(:mem_cache_store, ['127.0.0.1:11211'])

    cache.write("key1", true)
    cache.write("key2", true)
    cache.write("key3", true)

    fetch_multi_time = Benchmark.measure do
      10_000.times do
        cache.fetch_multi("key1", "key2", "key3") do |_key|
          raise 'miss'
        end
      end
    end

    read_write_time = Benchmark.measure do
      10_000.times do
        cache.read("key1") or raise 'miss'
        cache.read("key2") or raise 'miss'
        cache.read("key3") or raise 'miss'
      end
    end

    render json: {
      fetch_multi_time: fetch_multi_time.real,
      read_write_time: read_write_time.real
    }
  end
end

require "minitest/autorun"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_returns_success
    get "/"
    assert last_response.ok?
    raise last_response.body
  end

  private
    def app
      Rails.application
    end
end
```

You have to create a real application, add `dalli` to Gemfile, configure the environment to use mem cache as cache store and create a controller with the same code.

```ruby
class MoviesController < ApplicationController
  def index
    Rails.cache.write("key1", true)
    Rails.cache.write("key2", true)
    Rails.cache.write("key3", true)

    fetch_multi_time = Benchmark.measure do
      1_000.times do
        Rails.cache.fetch_multi("key1", "key2", "key3") do |_key|
          raise 'miss'
        end
      end
    end

    read_write_time = Benchmark.measure do
      1_000.times do
        Rails.cache.read("key1") or raise 'miss'
        Rails.cache.read("key2") or raise 'miss'
        Rails.cache.read("key3") or raise 'miss'
      end
    end

    render json: {
      fetch_multi_time: fetch_multi_time.real,
      read_write_time: read_write_time.real
    }
  end
end
```

### Expected behavior

I was expecting the `fetch_multi` to perform a lot better than using 3 separated reads.

### Actual behavior

Be aware that using the bug report template does not reproduce the issue (work as expected):

```js
{
  "fetch_multi_time": 2.637000000104308,
  "read_write_time": 3.379054999910295
}
```

But using a fresh application does reproduce (work as not expected):

```javascript
{
  "fetch_multi_time": 0.3341950001195073,
  "read_write_time": 0.01899300003424287
}
```

read + write time is an order of magnitude faster than using fetch multi and this is not supposed to be true.

Note that using dalli directly, the performance is slight better as expected (would be better by having network latency):

```ruby
  def index
    if params[:dalli]
      dalli = Dalli::Client.new('localhost:11211')
      dalli.set("key1", true)
      dalli.set("key2", true)
      dalli.set("key3", true)

      fetch_multi_time = Benchmark.measure do
        1_000.times do
          dalli.get_multi("key1", "key2", "key3")
        end
      end

      read_write_time = Benchmark.measure do
        1_000.times do
          dalli.get("key1")
          dalli.get("key2")
          dalli.get("key3")
        end
      end
    else
      Rails.cache.write("key1", true)
      Rails.cache.write("key2", true)
      Rails.cache.write("key3", true)

      fetch_multi_time = Benchmark.measure do
        1_000.times do
          Rails.cache.fetch_multi("key1", "key2", "key3") do |_key|
            raise 'miss'
          end
        end
      end

      read_write_time = Benchmark.measure do
        1_000.times do
          Rails.cache.read("key1") or raise 'miss'
          Rails.cache.read("key2") or raise 'miss'
          Rails.cache.read("key3") or raise 'miss'
        end
      end
    end

    render json: {
      fetch_multi_time: fetch_multi_time.real,
      read_write_time: read_write_time.real
    }
  end
```

```js
{
  "fetch_multi_time": 0.23557999986223876,
  "read_write_time": 0.29412800003774464
}
```

So, by that we can assume that the problem is not the Memcached or Dalli driver.

These values are consistent on my machine across several reloads.

We may have some sort of optimization on read calls that aren't present on fetch_multi or something like that, I'm not sure.

### System configuration

**Rails version**: 5.1.6

**Ruby version**: 2.4.2

### PR DESCRIPTION
**Pull request by sobrinho**
### Summary

Cache stores have a local cache strategy that is not used on fetch multi because read multi entries is not aware of it.

Closes #31909.

### Other Information

Using bug report template won't detect the issue for some reason.

You have to create a real application, add dalli to Gemfile, configure the environment to use mem cache as cache store and create a controller with that:

```ruby
class MoviesController < ApplicationController
  def index
    if params[:dalli]
      d = Dalli::Client.new('127.0.0.1:11211')
      k = "x" * 1024

      if params[:clear]
        d.flush
      else
        d.set("#{k}1", true)
        d.set("#{k}2", true)
        d.set("#{k}3", true)
      end

      fetch_multi_time = Benchmark.measure do
        1_000.times do
          d.get_multi("#{k}1", "#{k}2", "#{k}3") { |_key| true }
        end
      end

      if params[:clear]
        d.flush
      else
        d.set("#{k}1", true)
        d.set("#{k}2", true)
        d.set("#{k}3", true)
      end

      read_write_time = Benchmark.measure do
        1_000.times do
          d.get("#{k}1") || d.set("#{k}1", true)
          d.get("#{k}2") || d.set("#{k}2", true)
          d.get("#{k}3") || d.set("#{k}3", true)
        end
      end

      render json: {
        fetch_multi_time: fetch_multi_time.real,
        read_write_time: read_write_time.real,
      }
    else
      driver = Rails.cache

      if params[:clear]
        driver.clear
      else
        driver.write("#{k}1", true)
        driver.write("#{k}2", true)
        driver.write("#{k}3", true)
      end

      fetch_multi_time = Benchmark.measure do
        1_000.times do
          driver.read_multi("#{k}1", "#{k}2", "#{k}3") { |_key| true }
        end
      end

      if params[:clear]
        driver.clear
      else
        driver.write("#{k}1", true)
        driver.write("#{k}2", true)
        driver.write("#{k}3", true)
      end

      read_write_time = Benchmark.measure do
        1_000.times do
          driver.read("#{k}1") || driver.write("#{k}1", true)
          driver.read("#{k}2") || driver.write("#{k}2", true)
          driver.read("#{k}3") || driver.write("#{k}3", true)
        end
      end

      render json: {
        fetch_multi_time: fetch_multi_time.real,
        read_write_time: read_write_time.real
      }
    end
  end
end
```

This benchmark is rudimentary but it shows the problem in the most simple way.

Before:

```js
{
  "fetch_multi_time": 1.5901559998746961,
  "read_write_time": 0.020418000174686313
}
```

After (the difference falls in no statistic difference):

```js
{
  "fetch_multi_time": 0.014100000029429793,
  "read_write_time": 0.023772000102326274
}
```

This was detected while writing some benchmarks for [abstract builder](https://github.com/sobrinho/abstract_builder).

### CODE REVIEW

**Comment by sobrinho** [comment]
I'm looking for code review on this one, the code is not ready yet.

cc @fatkodima

**Comment by fatkodima** [comment]
`RedisCacheStore` has this problem as well. Probably, duplicated module `LocalCacheWithRaw` should be extracted into parent `Store` and reused for both memcached and redis?

**Comment by sobrinho** [comment]
Good point, I will do that.

**Comment by sobrinho** [comment]
@fatkodima can you take a look again?

**Comment by sobrinho** [comment]
Rebased to fix the build.

File: activesupport/lib/active_support/cache/strategy/local_cache.rb [hunk 1]
```diff
@@ -116,6 +116,19 @@ def read_entry(key, options)
             end
           end
 
+          def read_multi_entries(names, options)
+            return super unless local_cache
+
+            local_entries = Hash[names.map { |name| [name, local_cache.read_entry(name, options)] }.keep_if { |_name, value| value }]
+            missed_keys = names - local_entries.keys
+
+            if missed_keys.present?
```
Comment (jeremy) [inline activesupport/lib/active_support/cache/strategy/local_cache.rb]: `missed_keys` will always be an array, so we can narrow this to `.any?`

File: activesupport/lib/active_support/cache/strategy/local_cache.rb [hunk 2]
```diff
@@ -116,6 +116,19 @@ def read_entry(key, options)
             end
           end
 
+          def read_multi_entries(names, options)
+            return super unless local_cache
+
+            local_entries = Hash[names.map { |name| [name, local_cache.read_entry(name, options)] }.keep_if { |_name, value| value }]
```
Comment (jeremy) [inline activesupport/lib/active_support/cache/strategy/local_cache.rb]: Let's extract to `LocalStore#read_multi_entries`

**Review by jeremy — CHANGES_REQUESTED** [review]

File: activesupport/lib/active_support/cache/strategy/local_cache.rb (hunk 2 above)
Comment (sobrinho) [reply activesupport/lib/active_support/cache/strategy/local_cache.rb]: Extract this line to another method?

This is the name of this method itself.

**Pull request closed without merging**